File: Blind75/TwoSum/TwoSum.py
# 1. Two Sum
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
# You can return the answer in any order.

# Example 1:
# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

# Example 2:
# Input: nums = [3,2,4], target = 6
# Output: [1,2]

# Example 3:
# Input: nums = [3,3], target = 6
# Output: [0,1]

# * My solution:
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Create a hashmap that we will be adding to and checking as we go
        # Key: Number, Value: Index
        seenNumbers = {}

        # Loop through the array (nums)
        for idx, number in enumerate(nums):
            # Calculate the difference between the target 
            # and the current number
            difference = target - number
            
            # If the difference already exists in the hashmap
            if difference in seenNumbers:
                # Return the indices of both numbers that
                # sum up to the target
                return [seenNumbers[difference], idx]
            
            # Otherwise, add the current number to the hashmap 
            # using the number as the key, index is the value of the number
            seenNumbers[number] = idx

# Test Cases
solution = Solution()

# Example 1, expected output = [0,1]
print(solution.twoSum([2,7,11,15], 9))

# Example 2, expected output = [1,2]
print(solution.twoSum([3,2,4], 6))

# Example 3, expected output = [0,1]
print(solution.twoSum([3,3], 6))

# Example 4, expected output = [0,7]
print(solution.twoSum([4,9,5,7,2,19,42,6], 10))

File: Blind75/TwoSum/TwoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    console.log(`\n => Checking to see which two nums from [${nums}] will sum to make (${target})`);

    // Create a hashmap to keep track of seen numbers and their indices
    const hashMap = {};

    // Loop through the array of numbers
    for (let i = 0; i < nums.length; i++) {
        // Current number
        let n = nums[i];

        // Calculate the difference needed to reach the target
        let difference = target - n;

        // Log current iteration details
        console.log(`Checking number at index ${i}: ${n}, Difference needed: ${difference}`);
        console.log(hashMap)

        // Check if the difference is already in the hashmap
        if (hashMap[difference] !== undefined) {
            // If yes, log and return the indices of the two numbers
            console.log(`Found a pair: ${difference} and ${n}`);
            return [hashMap[difference], i];
        } else {
            // If no: add the index of the current number to the hashmap
            console.log(`Adding ${n} at index ${i} to the hashmap`);
            hashMap[n] = i;
        }
    }
}

// Test Cases
let nums = [2, 7, 11, 15];
let target = 9;
console.log(twoSum(nums, target)); // [0, 1]

nums = [3, 2, 4];
target = 6;
console.log(twoSum(nums, target)); // [1, 2]

nums = [3, 3];
target = 6;
console.log(twoSum(nums, target)); // [0, 1]

nums = [3, 2, 3];
target = 6;
console.log(twoSum(nums, target)); // [0, 2]